Lib/asyncio/selector_events.py

```diff
@@ -42,7 +42,7 @@ def _test_selector_event(selector, fd, event):
 if hasattr(socket, 'TCP_NODELAY'):
     def _set_nodelay(sock):
         if (sock.family in {socket.AF_INET, socket.AF_INET6} and
-                sock.type == socket.SOCK_STREAM and
+                base_events._is_stream_socket(sock.type) and
```

Comment:
At this point, how about switching to EAFP (Easier to Ask Forgiveness than Permission)?
```
try:
    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
except OSError:
    pass
```

Reply:
Yeah, that's what I'm planning to do here.  There appears to be no reliable **and** future-proof way of checking socket type on Linux, so EAFP is the best option.  The "not so good" option is to do a dance like this:

```python
if sock.type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC) == SOCK_STREAM:
```

Reply:
I prefer to filter flags rather than always calling setsockopt() but ignore errors. For best performance, I prefer to avoid useless syscalls.

> sock.type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)

Maybe we should add an attribute to socket.socket which returns the type without SOCK_NONBLOCK and SOCK_CLOEXEC bits.

Reply:
Another option is to add `SOCK_TYPE_MASK`, maybe even as `socket.SOCK_TYPE_MASK`, as Linux does (https://github.com/torvalds/linux/blob/v4.13/include/linux/net.h#L77).

I think that
```python
sock.type & ~(SOCK_NONBLOCK | SOCK_CLOEXEC)
```
is not as future-proof as
```python
sock.type & SOCK_TYPE_MASK
```
Shouldn't BSD/Linux devs support backward-compatibility/conventions after introducing `SOCK_CLOEXEC ` and `SOCK_NONBLOCK ` flags? Maybe using `SOCK_TYPE_MASK` will be safe enough?

Reply:
I like `SOCK_TYPE_MASK` idea.  At least we can do this on Linux.

Reply:
I'll do some experiments/more digging and update this PR.

Reply:
> For best performance, I prefer to avoid useless syscalls.

At the cost of maintainability, for a syscall that's called once in a socket's lifetime...

Reply:
About being "future-proof": SOCK_NONBLOCK and SOCK_CLOEXEC flags are now old and well known. But I agree that it's annoying to have to care of them :-(

Reply:
```python
>>> def f(sock):
...:    try:
...:        sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
...:    except OSError:
...:        pass
...:    
>>> sock = socket.socket(socket.AF_INET)
>>> %timeit f(sock)
500 ns ± 18.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
>>> sock = socket.socket(socket.AF_UNIX)
>>> %timeit f(sock)
1.66 µs ± 6.82 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```